libs/core/langchain_core/messages/utils.py

```diff
@@ -242,3 +248,604 @@ def convert_to_messages(
         List of messages (BaseMessages).
     """
     return [_convert_to_message(m) for m in messages]
+
+
+def filter_messages(
+    messages: Sequence[MessageLikeRepresentation],
+    *,
+    incl_names: Optional[Sequence[str]] = None,
+    excl_names: Optional[Sequence[str]] = None,
+    incl_types: Optional[Sequence[Union[str, Type[BaseMessage]]]] = None,
+    excl_types: Optional[Sequence[Union[str, Type[BaseMessage]]]] = None,
+    incl_ids: Optional[Sequence[str]] = None,
+    excl_ids: Optional[Sequence[str]] = None,
+) -> List[BaseMessage]:
+    """Filter messages based on name, type or id.
+
+    Args:
+        messages: Sequence Message-like objects to filter.
+        incl_names: Message names to include.
+        excl_names: Messages names to exclude.
+        incl_types: Message types to include. Can be specified as string names (e.g.
+            "system", "human", "ai", ...) or as BaseMessage classes (e.g.
+            SystemMessage, HumanMessage, AIMessage, ...).
+        excl_types: Message types to exclude. Can be specified as string names (e.g.
+            "system", "human", "ai", ...) or as BaseMessage classes (e.g.
+            SystemMessage, HumanMessage, AIMessage, ...).
+        incl_ids: Message IDs to include.
+        excl_ids: Message IDs to exclude.
+
+    Returns:
+        A list of Messages that meets at least one of the incl_* conditions and none
+        of the excl_* conditions. If not incl_* conditions are specified then
+        anything that is not explicitly excluded will be included.
+
+    Raises:
+        ValueError if two incompatible arguments are provided.
+
+    Example:
+        .. code-block:: python
+
+            from langchain_core.messages import filter_messages, AIMessage, HumanMessage, SystemMessage
+
+            messages = [
+                SystemMessage("you're a good assistant."),
+                HumanMessage("what's your name", id="foo", name="example_user"),
+                AIMessage("steve-o", id="bar", name="example_assistant"),
+                HumanMessage("what's your favorite color", id="baz",),
+                AIMessage("silicon blue", id="blah",),
+            ]
+
+            filter_messages(
+                messages,
+                incl_names=("example_user", "example_assistant"),
+                incl_types=("system",),
+                excl_ids=("bar",),
+            )
+
+        .. code-block:: python
+
+            [
+                SystemMessage("you're a good assistant."),
+                HumanMessage("what's your name", id="foo", name="example_user"),
+            ]
+    """  # noqa: E501
+    messages = convert_to_messages(messages)
+    filtered: List[BaseMessage] = []
+    for msg in messages:
+        if excl_names and msg.name in excl_names:
```

Comment:
where there are conflicts, here we prefer exclusions generally. would it make sense to instead prefer id > name > type? e.g., so I can say "exclude all human messages but include id 123, which is a human message"

Reply:
imo thats a bit too hard to grok as an end user